presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftHiveMetastore.java

```diff
@@ -831,7 +878,45 @@ public void dropPartition(String databaseName, String tableName, List<String> pa
     }
 
     @Override
-    public void alterPartition(String databaseName, String tableName, Partition partition)
+    public void alterPartition(String databaseName, String tableName, PartitionWithStatistics partitionWithStatistics)
+    {
+        Partition oldPartition = getPartition(databaseName, tableName, partitionWithStatistics.getPartition().getValues())
+                .orElseThrow(() -> new PartitionNotFoundException(new SchemaTableName(databaseName, tableName), partitionWithStatistics.getPartition().getValues()));
+        String partitionName = partitionWithStatistics.getPartitionName();
+        List<String> dataColumns = oldPartition.getSd().getCols().stream()
+                .map(FieldSchema::getName)
+                .collect(toImmutableList());
+        List<ColumnStatisticsObj> oldStatistics = getMetastorePartitionColumnStatistics(databaseName, tableName, ImmutableSet.of(partitionName), dataColumns)
+                .getOrDefault(partitionName, ImmutableList.of());
+
+        alterMetastorePartition(databaseName, tableName, toMetastoreApiPartition(partitionWithStatistics));
+
+        try {
+            // remove statistics for the columns that will not be overwritten
+            Set<String> columnStatisticsForRemoval = oldStatistics.stream()
+                    .map(ColumnStatisticsObj::getColName)
+                    .collect(toSet());
+            columnStatisticsForRemoval.removeAll(partitionWithStatistics.getStatistics().getColumnStatistics().keySet());
+            for (String columnName : columnStatisticsForRemoval) {
+                deletePartitionColumnStatistics(databaseName, tableName, partitionName, columnName);
+            }
+            storePartitionColumnStatistics(databaseName, tableName, partitionName, partitionWithStatistics);
+        }
+        catch (Throwable t) {
+            try {
+                alterMetastorePartition(databaseName, tableName, oldPartition);
```

Comment:
> For addPartitions, what we do is we refetch and check the embedded query id. We check every individual partition, instead of just one (by depending on atomicity). However, I don't know whether it's possible to do that for stats.

Yes, you are right. Removing the compensation actions in `addPartitions`

> For alterPartitions, we unconditionally undo everything. But even if we don't, we would have to undo the ones that failed because you couldn't know whether it succeeded remotely. When we do that, we'll need a mechanism to tag the stats we published.

We did this conditionally if the `alterPartition` call succeeded. Changed it to be unconditional, and removed the compensation actions from the `alterPartition`. (In fact it is still conditional. But instead of not rolling back if the call didn't succeed, now it won't be trying to roll back if we never tried to run the alter action)

> I can see this being a requirement for addPartitions. But what does this even mean for alterPartition? I don't understand what atomicity means when there's only one partition. (Then again, I don't think this matters much. See the next paragraph. We don't actually depend on the atomicity of addPartitions.)

I removed the comment about atomicity from the `addPartitions` methods in `HiveMetastore` and in `ExtendedHiveMetastore`

Reply:
> To be clear, I'm not saying SemiTransactionalHiveMetastore is the only way to do this. And I'm absolutely open to alternatives.

I also think that the `SemiTransactionalHiveMetastore` is the best place to do the compensation actions.

> What about stats that were dropped or altered? Do we want to roll back those?

Since we are going to do the rollback action in the `SemiTransactionalHiveMetastore`, it will rollback the old statistics automatically.

> Refetching the partition before every alter feels extremely inefficient.

I removed that. I'm going to remove the statistics based on the columns provided in the new partition object. The column set in the new partition objects reflects the current table columns. Although we may potentially leave some trash in the metastore, i don't think it is worth to explicitly remove all the statistics for the columns that are no longer in the table. These statistics won't be fetched any way.